drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,901 @@
+
+	if (fsl_cfg->masterSlave == kI2S_MasterSlaveNormalMaster ||
+		fsl_cfg->masterSlave == kI2S_MasterSlaveWsSyncMaster) {
+		fsl_cfg->divider = base_frequency /
+				   i2s_cfg->frame_clk_freq /
+				   fsl_cfg->frameLength;
+	}
+
+	/*
+	 * Set frame and bit clock polarity according to
+	 * inversion flags.
+	 */
+	switch (i2s_cfg->format & I2S_FMT_CLK_FORMAT_MASK) {
+	case I2S_FMT_CLK_NF_NB:
+		break;
+	case I2S_FMT_CLK_NF_IB:
+		fsl_cfg->sckPol = !fsl_cfg->sckPol;
+		break;
+	case I2S_FMT_CLK_IF_NB:
+		fsl_cfg->wsPol = !fsl_cfg->wsPol;
+		break;
+	case I2S_FMT_CLK_IF_IB:
+		fsl_cfg->sckPol = !fsl_cfg->sckPol;
+		fsl_cfg->wsPol = !fsl_cfg->wsPol;
+		break;
+	default:
+		LOG_ERR("Unsupported clocks polarity");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static struct i2s_config *i2s_mcux_config_get(const struct device *dev,
+				enum i2s_dir dir)
+{
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream;
+
+	if (dir == I2S_DIR_RX) {
+		stream = &dev_data->rx;
+	} else {
+		stream = &dev_data->tx;
+	}
+
+	if (stream->state == I2S_STATE_NOT_READY) {
+		return NULL;
+	}
+
+	return &stream->cfg;
+}
+
+static int i2s_mcux_configure(const struct device *dev, enum i2s_dir dir,
+			       struct i2s_config *i2s_cfg)
+{
+	const struct i2s_mcux_config *cfg = dev->config;
+	struct i2s_mcux_data *dev_data = dev->data;
+	I2S_Type *base = cfg->base;
+	struct stream *stream;
+	uint32_t base_frequency;
+	i2s_config_t fsl_cfg;
+	int result;
+	uint8_t data_len = 0;
+	uint8_t bytes_per_word = 0;
+
+	if (dir == I2S_DIR_RX) {
+		stream = &dev_data->rx;
+	} else if (dir == I2S_DIR_TX) {
+		stream = &dev_data->tx;
+	} else {
+		LOG_ERR("Either RX or TX direction must be selected");
+		return -EINVAL;
+	}
+
+	if (stream->state != I2S_STATE_NOT_READY &&
+	    stream->state != I2S_STATE_READY) {
+		LOG_ERR("invalid state");
+		return -EINVAL;
+	}
+
+	if (i2s_cfg->frame_clk_freq == 0U) {
+		stream->state = I2S_STATE_NOT_READY;
+		return 0;
+	}
+
+	/* Figure out function base clock */
+	if (clock_control_get_rate(cfg->clock_dev,
+				   cfg->clock_subsys, &base_frequency)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Validate the configuration by converting it to SDK
+	 * format.
+	 */
+	result = i2s_mcux_flexcomm_cfg_convert(base_frequency, dir, i2s_cfg,
+					       &fsl_cfg);
+	if (result != 0) {
+		return result;
+	}
+
+	/* Apply the configuration */
+	if (dir == I2S_DIR_RX) {
+		I2S_RxInit(cfg->base, &fsl_cfg);
+	} else {
+		I2S_TxInit(cfg->base, &fsl_cfg);
+	}
+
+	/* Data length in bits */
+	data_len = (uint8_t)(((base->CFG1 & I2S_CFG1_DATALEN_MASK) >>
+						I2S_CFG1_DATALEN_SHIFT) + 1U);
+
+	/* Convert to bytes */
+	bytes_per_word = (data_len + 7) / 8U;
+
+	/* if one channel is disabled, bytes_per_word should be 4U, user should
+	 * pay attention that when data length is shorter than 16,
+	 * the data format: left data put in 0-15 bit and right data should put in 16-31
+	 */
+	if (((base->CFG1 & I2S_CFG1_ONECHANNEL_MASK) == 0U)) {
+		bytes_per_word = 4U;
+	}
```

Comment:
Please reference below descriptions

![TransferWidth](https://user-images.githubusercontent.com/16884721/112066307-1b72ca00-8b34-11eb-9cbb-bf56b603bd76.png)

Reply:
When the user is passing data to the `i2s_write` function the memory block is supposed to have all the data tightly packed next to each other. I.e. In case of the standard I2S data format and 8-bit word the frame data should fit in 2 bytes. The next frame data will occupy adjacent 2 bytes.

This is not very well documented, but is mentioned in `struct i2s_config` description.

It looks like this I2S hardware module + DMA engine are not compatible with how the API was designed. They enforce different data layout.

Would you be able to discuss this issue during today's API meeting?

Reply:
Sure, I plan to attend the call. I am guessing the issue is that the DMA would transferring extra bytes at the end?

Reply:
I2S IP not friendly to 8 bit data, so in this case, software should convert the tightly pack data to the format that I2S required, that is

![Picture1](https://user-images.githubusercontent.com/16884721/112323522-db206280-8c7f-11eb-85aa-c5983e363578.png)

The 32-24 bits and 15-8 bits are left with empty.